src/librustc_mir/interpret/intern.rs

```diff
@@ -0,0 +1,231 @@
+use rustc_data_structures::fx::FxHashSet;
+
+use syntax::ast::Mutability;
+
+use super::{
+    ValueVisitor, MemoryKind, Pointer, AllocId, MPlaceTy, InterpError, Scalar,
+};
+use crate::const_eval::{CompileTimeInterpreter, CompileTimeEvalContext};
+
+struct InternVisitor<'rt, 'a: 'rt, 'mir: 'rt, 'tcx: 'a+'rt+'mir> {
+    /// previously encountered safe references
+    ref_tracking: &'rt mut RefTracking<MPlaceTy<'tcx>>,
+    ecx: &'rt mut CompileTimeEvalContext<'a, 'mir, 'tcx>,
+    mode: InternMode,
+    /// A list of all encountered relocations. After type-based interning, we traverse this list to
+    /// also intern allocations that are only referenced by a raw pointer or inside a union.
+    leftover_relocations: &'rt mut FxHashSet<AllocId>,
+}
+
+#[derive(Copy, Clone, Debug)]
+enum InternMode {
+    StaticMut,
+    Static {
+        frozen: bool,
+    },
+    Const {
+        frozen: bool,
+    },
+}
+
+/// Signalling data structure to ensure we don't recurse
+/// into the memory of other constants or statics
+struct IsStaticOrFn;
+
+impl<'rt, 'a, 'mir, 'tcx> InternVisitor<'rt, 'a, 'mir, 'tcx> {
+    fn intern(&mut self, ptr: Pointer) -> EvalResult<'tcx, Option<IsStaticOrFn>> {
+        trace!(
+            "InternVisitor::intern {:?}",
+            ptr,
+        );
+        // remove allocation
+        let tcx = self.ecx.tcx;
+        let memory = self.ecx.memory_mut();
+        let (kind, mut alloc) = match memory.alloc_map.remove(&ptr.alloc_id) {
+            Some(entry) => entry,
+            None => {
+                // if the pointer is dangling (neither in local nor global memory), we leave it
+                // to validation to error. The `delay_span_bug` ensures that we don't forget such
+                // a check in validation.
+                if tcx.alloc_map.lock().get(ptr.alloc_id).is_none() {
+                    tcx.sess.delay_span_bug(self.ecx.tcx.span, "tried to intern dangling pointer");
+                }
+                // treat dangling pointers like other statics
+                // just to stop trying to recurse into them
+                return Ok(Some(IsStaticOrFn));
+            },
+        };
+        // This match is just a canary for future changes to `MemoryKind`, which most likely need
+        // changes in this function.
+        match kind {
+            MemoryKind::Stack | MemoryKind::Vtable => {},
+        }
+        // Ensure llvm knows not to put this into immutable memory if the value is guaranteed
+        // immutable by looking at its value's type and whether it's a constant or a static.
+        alloc.mutability = match self.mode {
+            InternMode::StaticMut => Mutability::Mutable,
+            InternMode::Static { frozen: true } => Mutability::Immutable,
+            InternMode::Static { frozen: ref mut frozen @ false } => {
+                // Values behind the initial reference are frozen (except if the type undoes this
+                // via `UnsafeCell`).
```

Comment:
Where does that "undoing" happen? I only see one call to `is_freeze` and that is on the outside, when computing the initial mode.

Reply:
Right now this doesn't happen at all, because you can't create such values (statically prevented)

Reply:
So where *would* it happen then? Or is that not implemented yet? If so, there should be a FIXME here saying that -- and ideally we'd make sure we run into an ICE instead of doing wrong codegen when that inevitably happens.

Reply:
It would happen in `visit_aggregate` if said aggregate is `UnsafeCell`. I added a check that if we get there, but are `frozen` (so behind an indirection), we ICE, because const qualif should have errored out already.